File: python/book-store/book_store.py
"""
Calculate the best discount available for a basket full of books in a given 5 book series
2 of the same in series is 5%, 3 is 10%, 4 is 20%, and all five gets you a 25% discount.
"""

SINGLE_BOOK = 800

DISCOUNT = {
    2: 0.05,
    3: 0.10,
    4: 0.20,
    5: 0.25,
}

COMBINATIONS = {

    5: [[1, 2, 3, 4, 5 ]],
    4: [[1, 2, 3, 4],
        [1, 2, 3, 5],
        [1, 2, 4, 5],
        [1, 3, 4, 5],
        [2, 3, 4, 5]],
    3: [[1, 2, 3],
        [1, 2, 4],
        [1, 2, 5],
        [1, 3, 4],
        [1, 3, 5],
        [1, 4, 5],
        [2, 3, 4],
        [2, 3, 5],
        [2, 4, 5],
        [3, 4, 5]],
    2: [[1, 2],
        [1, 3],
        [1, 4],
        [1, 5],
        [2, 3],
        [2, 4],
        [2, 5],
        [3, 4],
        [3, 5],
        [4, 5]]
}

CACHE = {}

def total(basket):
    """
    Compute the total after discount for books in a given basket (list of book numbers)
    """
    books_grouped = {1: 0, 2: 0, 3: 0, 4: 0, 5: 0}
    for book_num in basket:
        books_grouped[book_num] += 1
         
    return calc_discount(books_grouped) # default cost


def calc_discount(basket):
    """
    Return the smallest total for possible discounts for books in a basket.
    """
    min_total = sum(basket.values()) * SINGLE_BOOK
    unique_books = len(basket)

    for group_size in range(2, 6):
        if unique_books >= group_size:
            amount = check_groups(basket, group_size)
            min_total = min(amount, min_total)
    
    return min_total

def check_groups(basket, group_size):
    """
    Check the if a basket has an available discount for books of a given group size
    """
    min_total = float(sum(basket.values()) * SINGLE_BOOK)

    if group_size in COMBINATIONS:
        available = []
        for combination in COMBINATIONS[group_size]:
            is_available = True
            for value in combination:
                if value not in basket.keys():
                    is_available = False
                    break
            if is_available:
                available.append(combination)
            
        if len(available) > 0:
            for combination in available:
                current = {**basket}
                for value in combination:
                    current[value] = current[value] - 1
                    if current[value] <= 0:
                        del current[value]

                discount = DISCOUNT.get(group_size, 0.0)
                group_total = group_size * SINGLE_BOOK * (1.00 - discount)

                # Don't bother with a recursive call unless we are smaller than the current minimum
                print(f"Group Total: {group_total}")
                print(f"Min Total: {min_total}")
                if len(current) > 0 and group_total < min_total:
                    group_total += calc_discount_memoize(current)
                
                min_total = min(group_total, min_total)
    
    return min_total

def calc_discount_memoize(basket):
    """
    Calculate the discount but used the memoized value if possible to reduce runtime.
    """
    # It doesn't matter if we have 2,2,3 or 2,3,3 the discount will be the same.
    # so sort the counts ascending. Then turn that into a string so it will be keyed
    # on value not reference. Then cache/retrieve it so we don't have to
    # do it a million times. I had __HUGE__ runtimes before I got this working.

    key = ','.join([str(count) for count in sorted(basket.values())])
    if key in CACHE:
        return CACHE[key]
    ret = calc_discount(basket)
    CACHE[key] = ret
    return ret

File: csharp/book-store/BookStore.cs
public static class BookStore
{
    const decimal SINGLE_BOOK = 8.0M;
    static Dictionary<int, decimal> Discount = new Dictionary<int, decimal>()
    {
        [2] = 0.05M,
        [3] = 0.10M,
        [4] = 0.20M,
        [5] = 0.25M,
    };

    static Dictionary<int, List<List<int>>> allCombinations = new()
    {

        [5] = new List<List<int>>()
        {
            new() {1, 2, 3, 4, 5 },
        },
        [4] = new List<List<int>>()
        {
            new() {1, 2, 3, 4},
            new() {1, 2, 3, 5},
            new() {1, 2, 4, 5},
            new() {1, 3, 4, 5},
            new() {2, 3, 4, 5},
        },
        [3] = new List<List<int>>()
        {
            new() {1, 2, 3 },
            new() {1, 2, 4 },
            new() {1, 2, 5 },
            new() {1, 3, 4 },
            new() {1, 3, 5 },
            new() {1, 4, 5 },
            new() {2, 3, 4 },
            new() {2, 3, 5 },
            new() {2, 4, 5 },
            new() {3, 4, 5 },
        },
        [2] = new List<List<int>>()
        {
            new() {1, 2 },
            new() {1, 3 },
            new() {1, 4 },
            new() {1, 5 },
            new() {2, 3 },
            new() {2, 4 },
            new() {2, 5 },
            new() {3, 4 },
            new() {3, 5 },
            new() {4, 5 },
        }
    };

    static Dictionary<string, decimal> cache = new Dictionary<string, decimal>();

    public static decimal Total(IEnumerable<int> books)
    {
        Dictionary<int, int> booksGrouped = (from int book in books
                                             group book by book into pair
                                             select new KeyValuePair<int, int>(pair.Key, pair.Count())).ToDictionary<int, int>();

        return CalcDiscount(booksGrouped);
    }

    public static decimal CalcDiscount(Dictionary<int, int> books)
    {
        decimal minTotal;

        // Single book price for all as base case.
        minTotal = (from pair in books select pair.Value).Sum() * SINGLE_BOOK;

        for (int i = 2; i <= 5; i++)
        {
            if (books.Keys.Count >= i)
            {
                decimal total = CheckGroup(books, i);
                if (total < minTotal)
                    minTotal = total;
            }
        }

        return minTotal;
    }

    static decimal CheckGroup(Dictionary<int, int> books, int groupSize)
    {
        decimal minTotal = (from pair in books select pair.Value).Sum() * SINGLE_BOOK;        
        if (allCombinations.ContainsKey(groupSize))
        {
            List<List<int>> combinations = allCombinations[groupSize];

            var available = (from combination in combinations
                         where combination.All<int>(n => books.ContainsKey(n))
                         select combination);
            if (available != null)
            {
                foreach (List<int> combination in available)
                {
                    Dictionary<int, int> current = new(books);
                    foreach (int i in combination)
                    {
                        current[i] = current[i] - 1;
                        if (current[i] <= 0)
                            current.Remove(i);
                    }
                    if (!Discount.TryGetValue(groupSize, out decimal discount))
                        discount = 0.0M;
                    decimal total = groupSize * SINGLE_BOOK * (1.00M - discount);

                    // Don't bother with a recursive call unless we are smaller than the current minimum
                    if (current.Count > 0 && total < minTotal)
                        total += CalcDiscountMemoize(current);
                        
                    if (total < minTotal)
                            minTotal = total;
                }
            }
        }
        return minTotal;
    }

    public static decimal CalcDiscountMemoize(Dictionary<int, int> books)
    {
        // It doesn't matter if we have 2,2,3 or 2,3,3 the discount will be the same.
        // so sort the counts ascending. Then turn that into a string so it will be keyed
        // on value not reference. Then cache/retrieve it so we don't have to
        // do it a million times. I had __HUGE__ runtimes before I got this working.

        int[] keyArr = (from int value in books.Values orderby value select value).ToArray<int>();
        string key = String.Join(',', (from int value in keyArr select value.ToString()));
        if (cache.ContainsKey(key))
            return cache[key];
        decimal ret = CalcDiscount(books);
        cache.Add(key, ret);
        return ret;
    } 
}